EISW-88506 Extend G-API classification benchmark to work with different inference backends

sivanov-work opened the pull request:
extends classification benchmark demo to work with:
OpenVINO 2.0, ONNXRT DML, ONNXRT OV EP

Is supposed to work with openCV 4.9: https://github.com/opencv/opencv/pull/24060

sivanov-work commented:
@Wovchena @TolyaTalamanov 
Added backward compatible changes, expected integration with future ocv version: 4.9.0:
1) awaits  https://github.com/opencv/opencv/pull/24060
2) In order to be built, It requires for DirectML string ctor from the previous PR.

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
```diff
@@ -133,19 +162,12 @@ int main(int argc, char* argv[]) {
-        // clang-format on
-
+        execution_providers_t providers = createProvidersFromString(FLAGS_ep);
+        nets += util::create_execution_network<nets::Classification>(FLAGS_m, config, providers);
```
Why did you introduce `createProvidersFromString()` and `create_execution_network()` separately? Merging them would reduce the number of ifdefs

Wovchena replied:
Modularity has drawbacks. For example:
> [Complexity, coordination, testing, and integration are some of the common issues that arise.](https://www.linkedin.com/advice/0/what-benefits-challenges-modular-system-design)

The point i brought in the original comment refers to complexity. Can you explain me why you preferred modularity over complexity here?

sivanov-work replied:
I wouldn't like idea to discuss the pointless questions in review - like mixing up the concept of "System Architecture" (which reflected in the article) and the "Structured programming paradigm". 
By the way, according to the second paradigm - it'd better to have TWO functions: one for parsing arguments (like we have in multiple other demos here) and another one for using those arguments for creating functional entities.

Additionally, since we decided to introduce additional field 'backend' into arguments row it appeared that we are not required in MACRO anymore and such kind of "complexity" will gone

If you still do care about TWO functions instead of one I'm not concerned: let's unify it, because writing the demo, in my opinion, is not a place to carry out "profound architectural discussions"

Wovchena commented on demos/common/cpp_gapi/utils_gapi/include/gapi_features.hpp:
```diff
@@ -0,0 +1,8 @@
+#if CV_VERSION_MAJOR >= 4 && CV_VERSION_MINOR > 8
+    #define GAPI_IE_EXECUTION_PROVIDERS_AVAILABLE
+#endif
+
```
```suggestion
```

Wovchena commented on demos/common/cpp_gapi/utils_gapi/include/gapi_features.hpp:
Please move it to demos/common/cpp_gapi/utils_gapi/include/utils_gapi/gapi_features.hpp

Wovchena replied:
The problem is that there's `utils_gapi` in the path: `demos/common/cpp_gapi/utils_gapi/...` now. The style demos follow is that `include` subfolder layout should mimic upper folders layout, so writing `#include ""` results in mimicking the upper folders structure. If you want to emphasize
> that compile-time `features` must be more important in hierarchy than `utils`

you need to avoid mentioning `utils_gapi`in folder hierarchy.

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/classification_benchmark_demo_gapi.hpp:
How do I compile it? I made `#define GAPI_IE_EXECUTION_PROVIDERS_AVAILABLE` to be unconditional and run the following:
```bat
C:\Users\vzlobin\Downloads\w_openvino_toolkit_windows_2023.0.1.11005.fa1c41994f3_x86_64\w_openvino_toolkit_windows_2023.0.1.11005.fa1c41994f3_x86_64\setupvars.bat
git clone --recursive https://github.com/intel/onnxruntime.git
cd onnxruntime
.\build.bat --config Release --use_openvino CPU_FP32 --build_shared_lib --use_dml 
cd ..\opencv\build
git switch at/advanced-device-selection-onnxrt-directml
cmake -DWITH_INF_ENGINE=y -DWITH_ONNX=y -DORT_INCLUDE=C:\Users\vzlobin\r\onnxruntime\include -DONNXRT_ROOT_DIR=C:\Users\vzlobin\r\onnxruntime\build\Windows\packages\Microsoft.AI.DirectML.1.12.1 -DBUILD_TESTS=n -DBUILD_PERF_TESTS=n -DBUILD_EXAMPLES=n -DBUILD_opencv_apps=n .. && cmake --build . -j 50
cd C:\Users\vzlobin\r\z\demos\build
cmake .. -DOpenCV_DIR=C:\Users\vzlobin\r\opencv\build && cmake --build . --config Release -j
set PATH=C:\Users\vzlobin\r\opencv\build\bin\Release;%PATH%
.\intel64\Release\classification_benchmark_demo_gapi.exe -m C:\Users\vzlobin\.cache\omz\public\resnet-50-tf\FP16\resnet-50-tf.xml -i "C:\Users\vzlobin\OneDrive - Intel Corporation\a\test-data\ILSVRC2012_img_val" -labels C:\Users\vzlobin\r\z\data\dataset_classes\imagenet_2012.txt -ep OVEP:CPU
```
The demo fails with `[ ERROR ] G-API has been compiled without ONNX support`

Wovchena replied:
I've accounted for that by running `git switch at/advanced-device-selection-onnxrt-directml`. See the original comment. Is it possible to make it work given I'm willing to collect and recompile all custom brunches required for that?

sivanov-work replied:
Ok, I see - let me check it. It might be problem with building G-API with ONNX.
I just relied on the fact that G-API with ONNX linkage must be done without any problem, because the current PR doesn't affect that process

sivanov-work replied:
I've tried to repeat those steps and it seems that the instruction from here still works: https://github.com/opencv/opencv/wiki/Using-G-API-with-MS-ONNX-Runtime
it describes different parameters to set up onnx runtime for building gapi.
I managed to launch classification demo with ONNX/CPU backend successfully.

We could try DML/OVEP for ONNX later when official instructions goes on and when we update required version of openCV in this CMakeLists

Wovchena replied:
The demo refers to DirectML. As far as I understand it is only supported for Windows. I tried adapt the wiki you linked for Windows and did the following:
```bat
cd ..\..\onnxruntime
git clean -dfx
mkdir build
mkdir build/isntall
.\build.bat --config Release --build_shared_lib --parallel --use_openvino CPU_FP32 --use_dml --cmake_extra_defines CMAKE_INSTALL_PREFIX="C:\Users\vzlobin\r\onnxruntime\build\install"
cd build\Windows\Release
cmake --build . --target install
cd ..\..\..\..\opencv
git clean -dfx
mkdir build
cd build
cmake -DWITH_INF_ENGINE=y -DWITH_ONNX=y -DORT_INSTALL_DIR=C:/Users/vzlobin/r/onnxruntime/build/install -DBUILD_TESTS=n -DBUILD_PERF_TESTS=n -DBUILD_EXAMPLES=n -DBUILD_opencv_apps=n .. && cmake --build . --config Release -j
```
After the demo recompilation it still fails with the same error.

> We could try DML/OVEP for ONNX later when official instructions goes on and when we update required version of openCV in this CMakeLists

Do I read this correctly that there's no way to run it now? If so I don't see a point to continue looking at this until there's a way to make it work.

sivanov-work replied:
In my experiments I used Linux in order to build the appropriate openCV/GAPI with ONNX (without DMl/OVEP) to make sure that all code in demo compiled and  selected provider/backend are able to be applied. The goal of this experiment was to make sure that my dispatchers work as expected.

I didn't go further because I trust in https://github.com/opencv/opencv/pull/24060 works, since it passed local CI.
@TolyaTalamanov as the author of ONNX/{DML,OVEP} may be it's time for us to test it all together, please help us to establish set up for testing

TolyaTalamanov replied:
Hi @Wovchena, if you still encounter this problem:
```
[ ERROR ] G-API has been compiled without ONNX support
```
You probably didn't path `-DWITH_ONNX=ON` (I see you use `-DWITH_ONNX=y` instead)

After the build you should be able to run demo by using `ONNX_CPU` device.
If you want to run demo with particular `EP` (`DML`, `OVEP`, `TensorRT`, etc) you have to build `onnxrt` with that execution provider, for that please follow the official onnxrt documentation: https://onnxruntime.ai/docs/execution-providers/

sivanov-work commented on demos/classification_benchmark_demo/cpp_gapi/classification_benchmark_demo_gapi.hpp:
```diff
@@ -26,7 +31,14 @@ static const char no_show_message[] = "Optional. Disable showing of processed im
-
+#ifdef GAPI_IE_EXECUTION_PROVIDERS_AVAILABLE
+static const std::string ep_message_str("Optional. List of inference execution providers: " +
+                                    merge(getSupportedEP(), ",") +
```
agree, I will remove the  extra suffix

sivanov-work commented on demos/classification_benchmark_demo/cpp_gapi/classification_benchmark_demo_gapi.hpp:
```diff
@@ -42,7 +54,7 @@ DEFINE_string(res, "1280x720", image_grid_resolution_message);
 DEFINE_uint32(time, std::numeric_limits<gflags::uint32>::max(), execution_time_message);
 DEFINE_string(u, "", utilization_monitors_message);
-
+DEFINE_string(ep, "", ep_message);
```
ok

sivanov-work commented on demos/classification_benchmark_demo/cpp_gapi/src/execution_provider.cpp:
```diff
@@ -0,0 +1,37 @@
+
+
+execution_providers_t createProvidersFromString(const std::string &str){
+#ifndef GAPI_IE_EXECUTION_PROVIDERS_AVAILABLE
```
not necessary, as we discussed with @TolyaTalamanov  - it makes sense to unify EPs & Backends in one argument which would be version agnostic

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/classification_benchmark_demo_gapi.hpp:
```diff
@@ -26,7 +31,14 @@ static const char no_show_message[] = "Optional. Disable showing of processed im
+#ifdef GAPI_IE_EXECUTION_PROVIDERS_AVAILABLE
+static const std::string ep_message_str("Optional. List of inference execution providers: " +
+                                    merge(getSupportedEP(), ",") +
+                                    " - separated by \";\". Each provider followed by its device. Use \":\" as <provider:device> separator");
```
I've got a reply only to the first part.

> Please use , instead of ; to stay consistent with the rest lists, like -d

is still applicable. I just realized that using the same separators as OpenVINO uses: `:`, `,` can lead to ambiguity in `-ep DML:CPU;OVEP:HETERO:GPU,CPU`. So instead you may want to use completely different separators. I'm OK with any of two solutions: using only `:`, `,`, or using completely different ones. But please, don't use a mix of this two approaches.

sivanov-work commented:
@Wovchena @TolyaTalamanov Could your review it one more time please?
Added backend selection. Added parameters processing for ONNX backends
taking into account various string separators i.e. from openvino device parsing, the full configuration row has formats:
```
-backend IE
-backend OV
-backend ONNX
-backend ONNX/CPU
-backend ONNX/DML/VPU,ONNX/OVEP/GPU:0,ONNX/CPU
```

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
`.\intel64\Release\classification_benchmark_demo_gapi.exe -m C:\Users\vzlobin\.cache\omz\public\resnet-50-pytorch\resnet-v1-50.onnx -i "C:\Users\vzlobin\OneDrive - Intel Corporation\a\test-data\ILSVRC2012_img_val" -labels C:\Users\vzlobin\r\z\data\dataset_classes\imagenet_2012.txt -gt "C:\Users\vzlobin\OneDrive - Intel Corporation\a\test-data\ILSVRC2012_img_val\ILSVRC2012_val.txt" -backend "ONNX/DML/Intel(R) Iris(R) Xe Graphics"` may randomly crash for me. I reported that to @TolyaTalamanov, just leaving it as a reminder for myself to verify the fix later

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/classification_benchmark_demo_gapi.hpp:
```diff
@@ -26,7 +29,9 @@ static const char no_show_message[] = "Optional. Disable showing of processed im
                                              "Default is -1 (infinite time).";
 static const char utilization_monitors_message[] = "Optional. List of monitors to show initially.";
-
+static const std::string backend_message_str("Optional. Specify an inference backend (the list of available backends: " +
```
```suggestion
static const std::string backend_message_str("Optional. See README.md for details. Specify an inference backend (the list of available backends: " +
```

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
````diff
@@ -161,6 +162,8 @@ For example, use the following command-line command to run the application:
                       -u CDM
 ```
 
+The inference backend is specified by -backend flag. Depends on openCV version there might be different backends available: `IE` starting from openCV 4.0, `ONNX` from 4.7.1 and `OV` from 4.8.
````
```suggestion
To let the demo find onnx libs copy `onnxruntime_providers_openvino.dll` from `lib` and `onnxruntime.dll`, `onnxruntime_providers_shared.dll` from `bin` dirs  of onnxruntume install to the folder with the demo executable.

The inference backend is specified by -backend flag. Depends on openCV version there might be different backends available: `IE` starting from openCV 4.0, `ONNX` from 4.7.1 and `OV` from 4.8. The possible device value for `"ONNX/DML/<device>` can be found by running `set OPENCV_LOG_LEVEL=DEBUG` and ruining the demo with `--backend ONNX/DML/`. Search for `Available DirectML adapters:`. For ` -backend ONNX/OVEP/<device>` the device is the same as the one selected for `--use_openvino` while compiling `onnxruntime`.
```

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
`Cpp Demos (Ubuntu 20.04)` fails with:
`20:57:07  In file included from /opt/home/sys_k8sworker/workspace/omz/precommit/cpp_demos_ubuntu20/src/demos/common/cpp_gapi/utils_gapi/src/onnx_execution_provider.cpp:5:
20:57:07  /opt/home/sys_k8sworker/workspace/omz/precommit/cpp_demos_ubuntu20/src/demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_execution_provider.hpp: In function ‘void {anonymous}::applyDMLProvider(ExecNetwork&, const ModelConfig&, const inference_backends_t&)’:
20:57:07  /opt/home/sys_k8sworker/workspace/omz/precommit/cpp_demos_ubuntu20/src/demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_execution_provider.hpp:55:49: error: ‘cv::gapi::onnx::ep’ has not been declared
20:57:07     55 |     net.cfgAddExecutionProvider(cv::gapi::onnx::ep::DirectML(ep_selected_device))
20:57:07        |                                                 ^~
20:57:07  /opt/home/sys_k8sworker/workspace/omz/precommit/cpp_demos_ubuntu20/src/demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_execution_provider.hpp: In function ‘void {anonymous}::applyOVEPProvider(ExecNetwork&, const ModelConfig&, const inference_backends_t&)’:
20:57:07  /opt/home/sys_k8sworker/workspace/omz/precommit/cpp_demos_ubuntu20/src/demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_execution_provider.hpp:66:49: error: ‘cv::gapi::onnx::ep’ has not been declared
20:57:07     66 |     net.cfgAddExecutionProvider(cv::gapi::onnx::ep::OpenVINO(ep_selected_device));
20:57:07        |                                                 ^~`

I think It uses opencv which comes from `apt install`

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
`Build MSBuild Demos (Windows 10)` fails with
`20:58:28  C:\Jenkins\workspace\omz\precommit\demos_windows10\src\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/backend_builder.hpp(8,10): fatal error C1083: Cannot open include file: 'opencv2/gapi/infer/ov.hpp': No such file or directory [C:\Jenkins\workspace\omz\precommit\demos_windows10\demo-build\classification_benchmark_demo\cpp_gapi\classification_benchmark_demo_gapi.vcxproj]`

It uses opencv from

`run(['curl', '-L', 'https://github.com/opencv/opencv/releases/download/4.7.0/opencv-4.7.0-windows.exe', '--output', 'opencv-4.7.0-windows.exe'])`

`run(['opencv-4.7.0-windows.exe', '-oopencv', '-y'])`

sivanov-work replied:
having double check, I found out that ONNX for CPU is available since 4.7.0, which is updated in gapi_features.hpp
OV is still available starting from 4.8. But compilation failed due to missing MACRO guard

sivanov-work commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
I just run into the same issue. It seems that  ONNX model `resnet-50-pytorch\resnet-v1-50.onnx` draws expected result with ONNX gapi provider and gives wrong predictions with others two.
I presume we need to use OV IR models with gapi IE & OV providers. I tried that and got predicted result

Wovchena replied:
It terns out ONNX backend applies division by 255. Apparently that's good enough for resnet-50 even though it's incorrect, the correct values are https://github.com/openvinotoolkit/open_model_zoo/blob/ad80fe5b77543c2758c048c657b325af6306567a/models/public/resnet-50-pytorch/model.yml#L47-L48

sivanov-work commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
```diff
@@ -145,6 +145,7 @@ Options:
     -no_show                  Optional. Disable showing of processed images.
     -time "<integer>"         Optional. Time in seconds to execute program. Default is -1 (infinite time).
     -u                        Optional. List of monitors to show initially.
+    -backend <string>         Optional. Specify an inference backend. The list of available backends depend on openCV version. Default value is IE.
```
Sorry, I don't understand your point.
There is an actual output from benchmark:

```
build_ocv_2/intel64/Debug/classification_benchmark_demo_gapi -h
[ INFO ] OpenVINO
[ INFO ] 	version: 2023.0.1
[ INFO ] 	build: 2023.0.1-11048-a6351294e77-HEAD

classification_benchmark_demo [OPTION]
Options:
   ....
    -backend                  Optional. Specify an inference backend (the list of available backends: IE,ONNX,OV)
```
what kind of misalignment do you mention about?

Wovchena replied:
You have `Default value is IE.` in README, but not in help print. You should also stay consistent about using `()` for `the list of available backends`.

Another point which didn't exist at the moment of writing the initial comment is `See README.md for details.`

sivanov-work replied:
```
-backend                  Optional. Specify an inference backend. The list of available backends: IE,ONNX,OV. Default value is IE. See README.md for details
```
Do you think this is ok?

Wovchena replied:
Yes, as long as it is aligned between the implementation and README

sivanov-work commented on demos/common/cpp_gapi/utils_gapi/include/utils_gapi/backend_description.hpp:
```diff
@@ -0,0 +1,24 @@
+    std::vector<std::string> properties;
+    static BackendDescription parseFromArgs(const std::string &arg, char sep = '/');
+private:
+    BackendDescription(const std::vector<std::string> &fields);
```
the same reasons as before: modulatity, single-responsibility, etc.
parsing may be changed (due to another `backend` string format), but construction is still preserved

sivanov-work commented:
@Wovchena don't you mind to take another look and emphasize what I need to fix and what are other obstacles preventing to merge that?

Wovchena commented on demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_backend.hpp:
```diff
@@ -0,0 +1,118 @@
+        }
+        net.cfgAddExecutionProvider(Provider(ep_selected_device));
+    }
+};
```
I can't compile it with enabled `#define GAPI_ONNX_BACKEND_EP_EXTENSION`. The error is
`C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(63,2): error C2977: '`anonymous-namespace'
::ProviderApplicator': too many template arguments [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_gapi.vcxproj]
C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(54,8): message : see declaration of '`anon
ymous-namespace'::ProviderApplicator' [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_gapi.vcxproj]
C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(76,2): error C2953: '`anonymous-namespace'
::ProviderApplicator': class template has already been defined [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_ga
pi.vcxproj]
C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(66): message : see declaration of '`anonym
ous-namespace'::ProviderApplicator' [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_gapi.vcxproj]`

sivanov-work replied:
sorry, my fault. 
I'll fix that

sivanov-work replied:
checked it with different MACROs, it must be OK now

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
```diff
@@ -145,6 +145,7 @@ Options:
     -no_show                  Optional. Disable showing of processed images.
     -time "<integer>"         Optional. Time in seconds to execute program. Default is -1 (infinite time).
     -u                        Optional. List of monitors to show initially.
+    -backend <string>         Optional. Specify an inference backend. The list of available backends depend on openCV version. Default value is IE.
```
```suggestion
    -backend <string>         Optional. Specify an inference backend. The list of available backends depends on openCV version. Default value is IE. See README.md for details
```

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
`.\intel64\Release\classification_benchmark_demo_gapi.exe -m C:\Users\vzlobin\.cache\omz\public\resnet-50-pytorch\resnet-v1-50.onnx -i "C:\Users\vzlobin\OneDrive - Intel Corporation\a\test-data\ILSVRC2012_img_val" -labels C:\Users\vzlobin\r\z\data\dataset_classes\imagenet_2012.txt -gt "C:\Users\vzlobin\OneDrive - Intel Corporation\a\test-data\ILSVRC2012_img_val\ILSVRC2012_val.txt" -backend "ONNX/DML/Intel(R) Iris(R) Xe Graphics"` may randomly crash for me. I reported that to @TolyaTalamanov, just leaving it as a reminder for myself to verify the fix later

Wovchena commented on demos/common/cpp_gapi/utils_gapi/include/utils_gapi/gapi_features.hpp:
```diff
@@ -0,0 +1,20 @@
+#endif
+
+#if CV_VERSION_MAJOR >= 4 && CV_VERSION_MINOR >= 7
+    #define GAPI_ONNX_BACKEND
```
The presence of ONNX backend depends not only on OpenCV version but also on cmake flags used to compile OpenCV. Can you check if OpenCV already provides a define which could indicate that?

TolyaTalamanov replied:
Inference backend (a.k.a `ie`/`ov`/`onnx` params) exist always regardless backend availability.
In case `OpenCV` isn't compiled with proper inference backend it will throw the error in runtime

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/main.cpp:
It is up to you:

Mean/scale normalization is usually fused into IR. Now you support ONNX which doesn't have that. You may want to add `--mean_values` and `--scale_values` demo flags similar to original C++ demo

TolyaTalamanov replied:
>>> Do I need to process mean_value and scale_values If any of them is not empty.
If `mean_values` / `scale_values` are already exposed to `CLI` then yes.

What's about other backends: OV, IE ? Do we need to support scale/mean for them?
Yes, look at this: https://github.com/opencv/opencv/blob/4.x/modules/gapi/include/opencv2/gapi/infer/ov.hpp#L629-L654

You may throw exception for `IE` since this wasn't exposed for this backend but it's fine because it's gonna be deprecated soon (Otherwise we will include this for `4.9.0` hopefully)

TolyaTalamanov replied:
> How is it supposed to look like?
> 
> ```
> if (mean_values .empty() && scale_values .empty())   // both are empty?
> params.cfgNormalize({false}); 
> ```
> 
> Do I need to process `mean_value` and `scale_values` If any of them is not empty , like https://github.com/opencv/opencv/blob/4.8.0/modules/gapi/include/opencv2/gapi/infer/onnx.hpp#L184 ?
> 
> What's about other backends: OV, IE ? Do we need to support scale/mean for them?

Yes, if no mean & scale provided disable automatic normalization in `onnx` backend

sivanov-work replied:
thanks for clarification, I will improve that

Wovchena commented on demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_backend.hpp:
```diff
@@ -0,0 +1,120 @@
+        {"",     &ProviderApplicator<ExecNetwork>::apply},
+#ifdef GAPI_ONNX_BACKEND_EP_EXTENSION
+        {"DML",  &ProviderApplicator<ExecNetwork, cv::gapi::onnx::ep::DirectML>::apply},
+        {"OVEP", &ProviderApplicator<ExecNetwork, cv::gapi::onnx::ep::OpenVINO>::apply}
```
You have `EP` suffix for OV, but not for DML. Stay consistent

sivanov-work replied:
ok, i'll remove it. as discussed this suffix is not important

TolyaTalamanov commented on demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_backend.hpp:
```diff
@@ -0,0 +1,120 @@
+    static void apply(ExecNetwork &, const ModelConfig &, const inference_backends_t &backends) {
+        static_assert(std::is_same<Provider, CPUProvider>::value, "Unsupported ONNX provider requested. Please add a partial specialization if required");
+        if (backends.size() > 1) {
+            throw std::runtime_error("ONNX CPU execution provider must be only provider in the provider list or "
```
`CPU` provider isn't even exposed to `onnx::Params` I'd not mentioned it there at all

TolyaTalamanov replied:
If user defines `OVEP,DML` order it means that everything that don't map on these two will be transferred on `CPU` automatically. Despite user can't specify `OVEP,CPU,DML` since `CPU` supports everything there will be nothing to offload on `DML` in this case.

sivanov-work replied:
`--backend ONNX/CPU` was added behind `--backend ONNX` by reason of consistency to align with `ONNX/DMl`, `ONNX-OVEP`. When I have typed `ONNX/CPU`  and it would have throw me an error that `CPU provider is not supported` - then I wonder what provider is implied by `ONNX`

> . Despite user can't specify OVEP,CPU,DML since CPU supports everything there will be nothing to offload on DML in this case.

The condition above checks such scenario and throws error, that such combinations do not make sense. 
The only one stuff is seemed to be fixed here is removing ` or be the last in order`
@TolyaTalamanov  do you agree?

TolyaTalamanov replied:
1. `--backend ONNX` - Is this valid? Will be used default `CPU` backend?
2. `--backend ONNX/CPU` - I assume just `CPU` EP
3. `--backend ONNX/CPU/CPU` - is that valid? (I hope no)
4. `--backend ONNX/OV/CPU` - how would you figure out if `CPU` is the device for `OV` but not the second execution provider?
5. `--backend ONNX/OV/CPU/DML` - what is that line mean? I hope it will throw since `CPU` can be in the middle if it's `EP` (but look at the question in `4`)

sivanov-work replied:
> --backend ONNX - Is this valid? Will be used default CPU backend? 

ONNX on CPU is used

> --backend ONNX/CPU - I assume just CPU EP

ONNX on CPU is used

> --backend ONNX/CPU/CPU - is that valid? (I hope no)

Please check on code: https://github.com/openvinotoolkit/open_model_zoo/pull/3853/files#diff-a56ed8974441e654093998a6da1c245515cc44c781788d9b69061efeeb9a4692R49
any additional number of arguments is forbidden for ONNX CPU provider

> --backend ONNX/OV/CPU - how would you figure out if CPU is the device for OV but not the second execution provider?

From code:
List of backends  are parsed using `,` separator
https://github.com/openvinotoolkit/open_model_zoo/pull/3853/files#diff-b78e9f5bf604e1f15cbaea07f569ff7e9fd050fa3fc1ea6beb64c2333d840e90R66

List of attributes belong to specific backend are parsed using `/` https://github.com/openvinotoolkit/open_model_zoo/pull/3853/files#diff-714e493f5d95803493777c17cde54eb9637fdbef2354af62c990fc86ef1be889R21

So, ONNX- is backend,  OV - is onnx provider, CPU - is a provide args meant device
> --backend ONNX/OV/CPU/DML

is impossible here

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
````diff
@@ -161,6 +162,10 @@ For example, use the following command-line command to run the application:
                       -u CDM
 ```
 
+To let the demo find onnx libs copy `onnxruntime_providers_openvino.dll` from `lib` and `onnxruntime.dll`, `onnxruntime_providers_shared.dll` from `bin` dirs  of onnxruntume install to the folder with the demo executable.
````
`onnxruntime_providers_openvino.dll` is already mentioned

`DirectML.dll` doesn't exist for me. Are you sure?

TolyaTalamanov replied:
`DirectML.dll` should be in binary folder if you built `DML` execution provider.

TolyaTalamanov replied:
Well, in fact `DirectML.dll` is also part of `C:\system32` but if it's custom it also should be placed near the binary

sivanov-work replied:
I suppose whether anyone has proficient knowledge how to build the custom `DirectML.dll` instead of using the default one then this experimentator is well informed how to deploy it. In my perspective, this comprehensive instructions are redundant

Wovchena replied:
The instructions are mostly for me in first place. Usually I forget such specifics in two weeks.

TolyaTalamanov commented on demos/classification_benchmark_demo/cpp_gapi/include/execution_providers.hpp:
```diff
@@ -0,0 +1,78 @@
+
+inline const std::initializer_list<execution_provider_t> getSupportedEP() {
+#ifdef GAPI_IE_EXECUTION_PROVIDERS_AVAILABLE
+    static const std::initializer_list<execution_provider_t> providers{"DML", "OVEP"};
```
I think the point was using `--backend ONNX/OV/CPU` instead of `--backend ONNX/OVEP/CPU` if so - I'm fine with that

TolyaTalamanov commented:
> @Wovchena @TolyaTalamanov Could your review it one more time please? Added backend selection. Added parameters processing for ONNX backends taking into account various string separators i.e. from openvino device parsing, the full configuration row has formats:
> 
> ```
> -backend IE
> -backend OV
> -backend ONNX
> -backend ONNX/CPU
> -backend ONNX/DML/VPU,ONNX/OVEP/GPU:0,ONNX/CPU
> ```

Are `-backend ONNX` and `-backend ONNX/CPU` have the same behaviour?

TolyaTalamanov commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
```diff
@@ -161,6 +162,10 @@ For example, use the following command-line command to run the application:
 
+To let the demo find onnx libs copy `onnxruntime_providers_openvino.dll` from `lib` and `onnxruntime.dll`, `onnxruntime_providers_shared.dll` from `bin` dirs  of onnxruntume install to the folder with the demo executable.
+
+The inference backend is specified by -backend flag. Depends on openCV version there might be different backends available: `IE` starting from openCV 4.0, `ONNX` from 4.7.1 and `OV` from 4.8. The possible device value for `"ONNX/DML/<device>` can be found using `set OPENCV_LOG_LEVEL=DEBUG` and running the demo with `-backend ONNX/DML/`. Search for `Available DirectML adapters:`. For `-backend ONN ...
```
Was it fixed?

dmatveev replied:
> IE backend was introduced in opencv 4.5.2 version.

Actually in 4.2.0... https://github.com/opencv/opencv/wiki/ChangeLog#version420

dmatveev replied:
And ONNX in 4.5.1: https://github.com/opencv/opencv/wiki/ChangeLog#version451

sivanov-work replied:
Will be added

TolyaTalamanov commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
```diff
@@ -161,6 +162,10 @@ For example, use the following command-line command to run the application:
 
+To let the demo find onnx libs copy `onnxruntime_providers_openvino.dll` from `lib` and `onnxruntime.dll`, `onnxruntime_providers_shared.dll` from `bin` dirs  of onnxruntume install to the folder with the demo executable.
+
+The inference backend is specified by -backend flag. Depends on openCV version there might be different backends available: `IE` starting from openCV 4.0, `ONNX` from 4.7.1 and `OV` from 4.8. The possible device value for `"ONNX/DML/<device>` can be found using `set OPENCV_LOG_LEVEL=DEBUG` and running the demo with `-backend ONNX/DML/`. Search for `Available DirectML adapters:`. For `-backend ONN ...
```
Was it fixed

sivanov-work commented:
@TolyaTalamanov , @Wovchena  is anything else need to be fixed?

Wovchena commented on demos/classification_benchmark_demo/cpp_gapi/README.md:
```diff
@@ -161,6 +164,10 @@ For example, use the following command-line command to run the application:
 
+To let the demo find onnx libs copy `onnxruntime_providers_openvino.dll` from `lib` and `onnxruntime.dll`, `onnxruntime_providers_shared.dll` from `bin` dirs  of onnxruntume install to the folder with the demo executable.
+
+The inference backend is specified by -backend flag. Depends on openCV version there might be different backends available: `IE` starting from openCV 4.5.2, `ONNX` from 4.7.1 and `OV` from 4.8. The possible device value for `"ONNX/DML/<device>` can be found using `set OPENCV_LOG_LEVEL=INFO` and running the demo with `-backend ONNX/DML/`. Search for `Available DirectML adapters:`. For `-backend ON ...
```
```suggestion
The inference backend is specified by -backend flag. Depends on openCV version there might be different backends available: `IE` starting from openCV 4.5.2, `ONNX` from 4.7.1 and `OV` from 4.8. The possible device value for `"ONNX/DML/<device>` can be found using `set OPENCV_LOG_LEVEL=INFO` and running the demo with `-backend ONNX/DML/`. Search for `Available DirectML adapters:`. For `-backend ONNX/OV/<device>` the device is the same as the one selected for `--use_openvino` while compiling `onnxruntime`.
```

TolyaTalamanov commented on demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_backend.hpp:
```diff
@@ -0,0 +1,126 @@
+            };
+
+        // check on normalization
+        if (config.mean_values.empty() && config.scale_values.empty()) {
```
What if they are not empty?

TolyaTalamanov reviewed (approved):
LGTM 👍
